### DISCUSSION TITLE
How do I remove degenerate faces in Polyhedron_3?

### DISCUSSION THREAD

**Discussion by ankofl** [CGAL/cgal#8544] [Q&A] [answered]
After an unsuccessful attempt to combine two Polyhedron_3 due to the fact that some faces of one wall lie on the border of the faces of the other grid, I expand the smaller one in each direction by a small distance, after which the merging operation is successful. But as a result, those very slight protruding edges appear
![image](https://github.com/user-attachments/assets/7da30630-1287-49ff-98f1-6d81f52f699d)
For a better demonstration, I have increased the offset of the protruding faces, so that it would be clear what I am talking about:
![image](https://github.com/user-attachments/assets/2475d11a-b447-4dde-b1ee-39d602b47ed8)
Is it possible to somehow collapse or otherwise get rid of such faces, so that the result would be a flat surface without protruding 0.00000001 faces ?

**Comment by MaelRL** [comment]
If you need to make your boundaries conforming, there are other tools such as snapping / stitching.

Could you please provide the "before" & "after" mesh, as well as your code?

**Reply by ankofl** [reply]
```cpp
typedef CGAL::Exact_predicates_exact_constructions_kernel K;
typedef K::Point_3                                P;
typedef K::Vector_3                               V;
typedef std::vector<std::size_t> Face;

typedef CGAL::Polyhedron_3<K> Mesh;
````

I tried to use:
```cpp
PMP::remove_almost_degenerate_faces(mesh);
```

But at the moment I don't know how to set the settings for what are considered "degenerate faces". 

By the way, could you tell us more about the methods <tools such as snapping / stitching> in your message?

[Archive with original off-files needed to union](https://github.com/user-attachments/files/17367133/Unite.zip)

[off-file of Polyhedron_3 where need remove degenerated faces](https://github.com/user-attachments/files/17367087/united.zip)

[Project, where im work](https://github.com/ankofl/Native)

**Reply by ankofl** [reply]
The truth is, the whole problem is that:
```cpp
PMP::corefine_and_compute_union
```
returns false when I try to merge two Mesh edges that match or touch each other here:
```cpp
inline void union_meshes(
	const std::string& path_meshes,
	const std::string& path_mesh_output) {

	std::vector<Mesh> meshes;
	if (!load_dir(path_meshes, meshes)) {
		return;
	}

	Mesh first = meshes[0];
	for (size_t i = 1; i < meshes.size(); i++) {
		Mesh result;
		if (PMP::corefine_and_compute_union(first, meshes[i], result)) {
			first = result;
		}
		else {
		       // problem here
		}
		std::cout << i + 1 << '/' << meshes.size() << std::endl;
	}

	remesh_planar(first, first);

	save_to(first, path_mesh_output);
}
```

Perhaps there is a better way to solve this, rather than slightly increasing one of the Meshes in all directions? In this case, the problem with "degenerate faces" will be solved by itself. 

Thanks!

**Reply by ankofl** [reply, marked as answer]
@MaelRL  It is probably possible to close this top and move the discussion to the appropriate branch [created earlier](https://github.com/CGAL/cgal/discussions/8404)
